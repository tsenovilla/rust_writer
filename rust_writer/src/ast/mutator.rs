// SPDX-License-Identifier: GPL-3.0

use crate::Error;
use std::fmt::Debug;
use syn::{visit_mut::VisitMut, File};

#[derive(Debug, Clone)]
pub struct EmptyMutator;

pub struct Mutator<'a, T: Debug + Clone, const N: usize> {
	pub mutated: [bool; N],
	pub mutator: &'a T,
}

pub trait ToMutate<'a, T: Debug + Clone, const N: usize> {
	fn to_mutate(self, mutator: &'a T) -> Mutator<'a, T, N>;
}

impl Default for Mutator<'_, EmptyMutator, 1> {
	fn default() -> Self {
		Self { mutated: [false], mutator: &EmptyMutator }
	}
}

impl<'a, T, const N: usize> Mutator<'a, T, N>
where
	T: Debug + Clone,
	Mutator<'a, T, N>: VisitMut,
{
	pub fn mutate(mut self, ast: &mut File) -> Result<(), Error> {
		self.visit_file_mut(ast);

		if self.mutated.iter().all(|&x| x) {
			Ok(())
		} else {
			Err(Error::Descriptive(format!("Cannot mutate using Mutator: {:?}", self.mutator)))
		}
	}
}
