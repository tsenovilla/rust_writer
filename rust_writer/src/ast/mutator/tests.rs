// SPDX-License-Identifier: GPL-3.0

use super::*;
use crate::ast::implementors::ItemToTrait;
use syn::{parse_quote, TraitItem};
use test_builder::TestBuilder;

#[test]
fn mutator_works_for_a_implementor() {
	TestBuilder::default().with_trait_ast().execute(|mut builder| {
		let item_to_trait: ItemToTrait =
			("MyTrait", TraitItem::Type(parse_quote! {type Type1: From<String>;})).into();

		let ast = builder.get_mut_ast_file("trait.rs").expect("This exists; qed;");

		let mut mutator = Mutator::default().to_mutate(&item_to_trait);
		assert!(mutator.mutate(ast).is_ok());

		assert!(mutator.mutated[0]);

		mutator.reset();

		assert!(!mutator.mutated[0]);
	});
}
