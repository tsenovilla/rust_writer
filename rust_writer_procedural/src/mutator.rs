// SPDX-License-Identifier: GPL-3.0

mod expand;
mod parse;

use crate::parse_attrs::MacroAttrs;
use proc_macro::TokenStream;
use syn::{parse_macro_input, ItemStruct};

pub(crate) fn mutator(attrs: TokenStream, item: TokenStream) -> TokenStream {
	let attrs_list = parse_macro_input!(attrs as MacroAttrs);
	let struct_def = parse_macro_input!(item as ItemStruct);

	match parse::MutatorDef::try_from(attrs_list, struct_def) {
		Ok(def) => expand::expand_mutator(def).into(),
		Err(err) => err.to_compile_error().into(),
	}
}

pub(crate) fn impl_mutator(item: TokenStream) -> TokenStream {
	let mut finished = item.clone();

	let struct_def = parse_macro_input!(item as ItemStruct);

	let generated: TokenStream = match parse::ImplMutatorDef::try_from(struct_def) {
		Ok(def) => expand::expand_impl_mutator(def).into(),
		Err(err) => err.to_compile_error().into(),
	};

	finished.extend(generated);
	finished
}
