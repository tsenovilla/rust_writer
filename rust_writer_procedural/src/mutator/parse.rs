// SPDX-License-Identifier: GPL-3.0

use crate::parse_attrs::{InnerAttr, MacroAttrs};
use syn::{ItemStruct, Path, Result};

pub(crate) struct MutatorDef {
	pub(crate) implementors: Vec<Path>,
	pub(crate) struct_: ItemStruct,
	pub(crate) unit_struct: bool,
	pub(crate) impl_from: bool,
}

impl MutatorDef {
	pub(crate) fn try_from(attrs: MacroAttrs, mut struct_: ItemStruct) -> Result<Self> {
		let (unit_struct, impl_from) = match attrs.validate_struct(&mut struct_)? {
			InnerAttr::Unit => (true, false),
			InnerAttr::NotUnit => (false, false),
			InnerAttr::ImplFrom => (true, true),
			InnerAttr::ImplFromNotUnit => (false, true),
		};

		let implementors: Vec<Path> = attrs.0.into_iter().collect();

		Ok(Self { implementors, struct_, unit_struct, impl_from })
	}
}
