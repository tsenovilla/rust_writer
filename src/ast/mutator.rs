// SPDX-License-Identifier: GPL-3.0

use crate::Error;
use std::fmt::Debug;
use syn::{visit_mut::VisitMut, File};

#[derive(Debug)]
pub struct EmptyMutator;

pub struct Mutator<T: Debug> {
	pub mutated: bool,
	pub mutator: T,
}

pub trait ToMutate<T: Debug> {
	fn to_mutate(self, mutator: T) -> Mutator<T>;
}

impl Default for Mutator<EmptyMutator> {
	fn default() -> Self {
		Self { mutated: false, mutator: EmptyMutator }
	}
}

impl<T> Mutator<T>
where
	T: Debug,
	Mutator<T>: VisitMut,
{
	pub fn mutate(mut self, ast: &mut File) -> Result<(), Error> {
		self.visit_file_mut(ast);

		if self.mutated {
			Ok(())
		} else {
			Err(Error::Descriptive(format!("Cannot mutate using Mutator: {:?}", self.mutator)))
		}
	}
}
