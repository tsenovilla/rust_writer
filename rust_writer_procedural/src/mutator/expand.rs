// SPDX-License-Identifier: GPL-3.0

use crate::mutator::parse::MutatorDef;
use proc_macro2::{Span, TokenStream};
use quote::{quote, ToTokens};
use syn::{
	parse_quote, punctuated::Punctuated, token::Brace, Field, Fields, FieldsNamed, GenericParam,
	Ident, Index, Lifetime, PathArguments, Token, Type,
};

pub(crate) fn expand_mutator(def: MutatorDef) -> TokenStream {
	let MutatorDef { implementors, mut struct_, unit_struct, impl_from } = def;

	let struct_vis = &struct_.vis;
	let struct_name = &struct_.ident;

	let mutator_lifetime: Lifetime = parse_quote! {'mutator};
	let one = Index::from(1);

	let mut implementor_generics = Vec::new();

	let implementors_count = Index::from(implementors.iter().count());

	let implementors_indexes: Vec<Index> =
		implementors.iter().enumerate().map(|(index, _)| Index::from(index)).collect();

	let implementors_struct_names: Vec<Ident> = implementors
		.iter()
		.map(|implementor| {
			let last_implementor_segment = implementor
				.segments
				.last()
				.expect("At this point, implementors are valid paths; qed;");

			if let PathArguments::AngleBracketed(ref generics) = last_implementor_segment.arguments
			{
				generics.args.iter().for_each(|argument| {
					let generic_param: GenericParam = parse_quote!(#argument);
					if !implementor_generics.contains(&generic_param) {
						implementor_generics.push(generic_param);
					}
				})
			}

			Ident::new(
				&last_implementor_segment.ident.to_string().to_lowercase(),
				Span::call_site(),
			)
		})
		.collect();

	struct_.generics.params.extend(implementor_generics);

	let generics_params = &struct_.generics.params;

	let where_clause = struct_
		.generics
		.where_clause
		.as_ref()
		.map(|clause| clause.into_token_stream())
		.unwrap_or(quote! {where});

	let new_struct_fields: Vec<Field> = implementors_struct_names
		.iter()
		.zip(implementors)
		.map(|(name, implementor)| parse_quote!(#struct_vis #name: #implementor))
		.collect();

	if unit_struct {
		let mut fields: Punctuated<Field, Token![,]> = Punctuated::new();
		fields.extend(new_struct_fields);
		struct_.fields =
			Fields::Named(FieldsNamed { brace_token: Brace::default(), named: fields });
	} else {
		if let Fields::Named(ref mut fields) = struct_.fields {
			fields.named.extend(new_struct_fields);
		}
	}

	let mut impl_from_block = quote! {};

	if impl_from {
		if let Fields::Named(FieldsNamed { ref named, .. }) = struct_.fields {
			let fields_names: Vec<&Ident> = named
				.iter()
				.map(|field| field.ident.as_ref().expect("Named field has ident; qed;"))
				.collect();

			let fields_types: Punctuated<&Type, Token![,]> =
				named.iter().map(|field| &field.ty).collect();
			let tuple_indexes: Vec<Index> =
				named.iter().enumerate().map(|(index, _)| Index::from(index)).collect();

			impl_from_block = quote! {
				impl<#generics_params> From<(#fields_types)> for #struct_name<#generics_params> #where_clause{
					fn from(tuple: (#fields_types)) -> Self{
						Self{ #(#fields_names: tuple.#tuple_indexes),* }
					}
				}
			}
		}
	}


	let impl_to_mutate = quote! {
		impl<#mutator_lifetime, #generics_params>
		rust_writer::ast::mutator::ToMutate<#mutator_lifetime, #struct_name<#generics_params>, #implementors_count>
		for rust_writer::ast::mutator::Mutator<'_, rust_writer::ast::mutator::EmptyMutator, #one>
	  #where_clause
	  {
			fn to_mutate(self, mutator: &#mutator_lifetime #struct_name<#generics_params>)
			->
			rust_writer::ast::mutator::Mutator<#mutator_lifetime, #struct_name<#generics_params>,#implementors_count> {
				rust_writer::ast::mutator::Mutator {
					mutated: [false; #implementors_count],
				  mutator
				}
			}
		}
	};

	let impl_visit_mut = quote! {
		impl<#mutator_lifetime, #generics_params>
		syn::visit_mut::VisitMut
		for rust_writer::ast::mutator::Mutator<#mutator_lifetime, #struct_name<#generics_params>, #implementors_count>
	  #where_clause
	  {
			fn visit_file_mut(&mut self, file: &mut syn::File){
			#(
				  let mut mutator = rust_writer::ast::mutator::Mutator::default()
					  .to_mutate(&self.mutator.#implementors_struct_names);
				  mutator.visit_file_mut(file);
				  self.mutated[#implementors_indexes] = mutator.mutated.iter().all(|&x| x);
			)*
		 }
		}
	};

	quote! {
		#[rust_writer_procedural::already_expanded]
		#[derive(Debug, Clone)]
		#struct_
		#impl_from_block
		#impl_to_mutate
	  #impl_visit_mut
	}
	.into()
}
