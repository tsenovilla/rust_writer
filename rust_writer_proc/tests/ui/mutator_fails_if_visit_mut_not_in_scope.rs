// SPDX-License-Identifier: GPL-3.0

use rust_writer::ast::mutator::ToMutate;
use rust_writer_proc::mutator;

#[mutator(rust_writer::ast::implementors::ItemToTrait<'a>, rust_writer::ast::implementors::ItemToImpl<'a>)]
struct SomeStruct;

fn main() {}
