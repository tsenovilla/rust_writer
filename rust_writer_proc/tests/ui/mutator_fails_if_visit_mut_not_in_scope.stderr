error[E0599]: no method named `visit_file_mut` found for struct `Mutator` in the current scope
 --> tests/ui/mutator_fails_if_visit_mut_not_in_scope.rs:6:1
  |
6 | #[mutator(rust_writer::ast::implementors::ItemToTrait<'a>, rust_writer::ast::implementors::ItemToImpl<'a>)]
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
 ::: $CARGO/syn-2.0.100/src/gen/visit_mut.rs
  |
  |     fn visit_file_mut(&mut self, i: &mut crate::File) {
  |        -------------- the method is available for `Mutator<'_, ItemToTrait<'_>, 1>` here
  |
  = help: items from traits can only be used if the trait is in scope
  = note: this error originates in the attribute macro `mutator` (in Nightly builds, run with -Z macro-backtrace for more info)
help: trait `VisitMut` which provides `visit_file_mut` is implemented but not in scope; perhaps you want to import it
  |
3   + use syn::visit_mut::VisitMut;
    |
help: there is a method `visit_field_value_mut` with a similar name
    |
6   | visit_field_value_mut
    |

error[E0599]: no method named `visit_file_mut` found for struct `Mutator` in the current scope
 --> tests/ui/mutator_fails_if_visit_mut_not_in_scope.rs:6:1
  |
6 | #[mutator(rust_writer::ast::implementors::ItemToTrait<'a>, rust_writer::ast::implementors::ItemToImpl<'a>)]
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
 ::: $CARGO/syn-2.0.100/src/gen/visit_mut.rs
  |
  |     fn visit_file_mut(&mut self, i: &mut crate::File) {
  |        -------------- the method is available for `Mutator<'_, ItemToImpl<'_>, 1>` here
  |
  = help: items from traits can only be used if the trait is in scope
  = note: this error originates in the attribute macro `mutator` (in Nightly builds, run with -Z macro-backtrace for more info)
help: trait `VisitMut` which provides `visit_file_mut` is implemented but not in scope; perhaps you want to import it
  |
3   + use syn::visit_mut::VisitMut;
    |
help: there is a method `visit_field_value_mut` with a similar name
    |
6   | visit_field_value_mut
    |
